internal/controller/thanosquery_controller.go

```diff
@@ -132,29 +140,116 @@ func (r *ThanosQueryReconciler) buildQuerier(query monitoringthanosiov1alpha1.Th
 		LogFormat: query.Spec.LogFormat,
 	}.ApplyDefaults()
 
+	endpoints := r.getStoreAPIServiceEndpoints(ctx, query)
 	return manifestquery.BuildQuerier(manifestquery.QuerierOptions{
 		Options:       metaOpts,
 		ReplicaLabels: query.Spec.QuerierReplicaLabels,
 		Timeout:       "15m",
 		LookbackDelta: "5m",
 		MaxConcurrent: 20,
-
-		// TODO(saswatamcode): Add logic to for automated tracking of endpoints.
-		Endpoints: []manifestquery.Endpoint{{
-			ServiceName: "thanos-receive",
-			Namespace:   query.GetNamespace(),
-			Type:        manifestquery.Regular,
-		}},
+		Endpoints:     endpoints,
 	})
 }
 
+// getStoreAPIServiceEndpoints returns the list of endpoints for the StoreAPI services that match the ThanosQuery storeLabelSelector.
+func (r *ThanosQueryReconciler) getStoreAPIServiceEndpoints(ctx context.Context, query monitoringthanosiov1alpha1.ThanosQuery) []manifestquery.Endpoint {
+	services := &corev1.ServiceList{}
+	listOpts := []client.ListOption{
+		client.MatchingLabels(query.Spec.StoreLabelSelector.MatchLabels),
+		client.InNamespace(query.Namespace),
+	}
+	if err := r.List(ctx, services, listOpts...); err != nil {
+		return []manifestquery.Endpoint{}
+	}
+
+	if len(services.Items) == 0 {
+		return []manifestquery.Endpoint{}
+	}
+
+	endpoints := make([]manifestquery.Endpoint, len(services.Items))
+	for i, svc := range services.Items {
+		etype := manifestquery.RegularLabel
+
+		if metav1.HasLabel(svc.ObjectMeta, string(manifestquery.StrictLabel)) {
+			etype = manifestquery.StrictLabel
+		} else if metav1.HasLabel(svc.ObjectMeta, string(manifestquery.GroupStrictLabel)) {
+			etype = manifestquery.GroupStrictLabel
+		} else if metav1.HasLabel(svc.ObjectMeta, string(manifestquery.GroupLabel)) {
+			etype = manifestquery.GroupLabel
+		}
+
+		for _, port := range svc.Spec.Ports {
+			if port.Name == "grpc" {
+				endpoints[i].Port = port.Port
+				break
+			}
```

Comment:
shouldnt we bomb out here if we dont match on port? we would just be adding an endpoint without a port or is that ok?

Reply:
This is only for group, can proceed